proposals/0000-macro-forward.md

```diff
@@ -0,0 +1,84 @@
+# Expression macro method forwarding
+
+* Proposal: [HXP-NNNN](NNNN-filename.md)
+* Author: [Dan Korostelev](https://github.com/nadako)
+
+## Introduction
+
+Provide a way to have expression macro functions inside a class without typing that whole class
+in macro context.
+
+## Motivation
+
+It's often desirable to have a macro function inside a "normal" class/abstract to achieve
+nice API.
+
+However, doing this will cause the module containing a macro function to be also typed
+in the macro context, even if the macro function doesn't use anything from that module.
+
+This in turn can lead to annoying and surprising errors, since the other code in that
+module wasn't written with macros in mind and can depend on APIs that aren't even available
+in macros (e.g. target-specific ones).
+
+For example:
+```haxe
+class Main {
+    static function main() {
+        js.Browser.console.log("Version " + getVersion());
+    }
+
+    static macro function getVersion() {
+        return MyMacroTools.getVersionExpr();
+    }
+}
+```
+This will fail with `You cannot access the js package while in a macro (for js.Browser)`
+without even specifying the actual cause of the error.
+
+And even if it does pass through the typer, most of the code will simply be
+unused in macro context, so it's just a needless work consuming precious time.
+
+The current solution is to fence the non-macro code (and imports) with `#if !macro`,
+which hurts readability and often confuses people not familiar with macros, causing
+them to "wtf" by accidentally adding a function outside of `#if !macro` region.
+
+What I'd like to propose is a way to avoid the very cause of the issue by forwarding
+macro calls into another module.
+
+## Detailed design
+
+The proposed solution is quite simple: allow macro methods without body if they have
+a special metadata pointing to the static method in another type. For example, the function
+from the example above could be defined like this instead:
+
+```haxe
+@:forwardMacro(MyMacroTools.getVersionExpr)
+static macro function getVersion();
```

Comment:
Because, as stated in the beginning, we often want macro methods to be members of our class, so we could call it like `Main.getVersion()`.